File: 2015/16/solution.java

import java.util.HashMap;
import java.util.Scanner;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class AoC_16 {

    void magic() {
        HashMap<String, Integer> correctSue = new HashMap<>();
        correctSue.put("children", 3);
        correctSue.put("cats", 7);
        correctSue.put("samoyeds", 2);
        correctSue.put("pomeranians", 3);
        correctSue.put("akitas", 0);
        correctSue.put("vizslas", 0);
        correctSue.put("goldfish", 5);
        correctSue.put("trees", 3);
        correctSue.put("cars", 2);
        correctSue.put("perfumes", 1);

        int partOne = 0, partTwo = 0;
        while (sc.hasNextLine() && (partOne == 0 || partTwo == 0)) {
            String input = sc.nextLine();
            boolean isPartOne = true, isPartTwo = true;
            for (Matcher matcher = Pattern.compile("([a-z]+): (-?\\d+)").matcher(input); matcher.find();) {
                isPartOne &= correctSue.get(matcher.group(1)) == Integer.valueOf(matcher.group(2));
                switch (matcher.group(1)) {
                    case "cats":
                    case "trees":
                        isPartTwo &= correctSue.get(matcher.group(1)) < Integer.valueOf(matcher.group(2));
                        break;
                    case "pomeranians":
                    case "goldfish":
                        isPartTwo &= correctSue.get(matcher.group(1)) > Integer.valueOf(matcher.group(2));
                        break;
                    default:
                        isPartTwo &= correctSue.get(matcher.group(1)) == Integer.valueOf(matcher.group(2));
                        break;
                }
            }
            if (isPartOne) {
                partOne = Integer.valueOf(input.split(" ")[1].substring(0, input.split(" ")[1].length() - 1));
            }
            if (isPartTwo) {
                partTwo = Integer.valueOf(input.split(" ")[1].substring(0, input.split(" ")[1].length() - 1));
            }
        }
        System.out.println("Part One: " + partOne);
        System.out.println("Part Two: " + partTwo);
    }

    static Scanner sc = new Scanner(System.in);

    public static void main(String[] args) {
        long startTime = System.currentTimeMillis();
        new AoC_16().magic();
        System.out.println("Running time: " + (System.currentTimeMillis() - startTime) + "ms");
    }
}

File: 2015/16/solution.php
<?php

const COMPOUNDS = [
    'children'    => 3,
    'cats'        => 7,
    'samoyeds'    => 2,
    'pomeranians' => 3,
    'akitas'      => 0,
    'vizslas'     => 0,
    'goldfish'    => 5,
    'trees'       => 3,
    'cars'        => 2,
    'perfumes'    => 1
];

function compare(string $name, int $test, bool $corrected = false): bool
{
    if ($corrected) {
        if (in_array($name, ['cats', 'trees'])) {
            return COMPOUNDS[$name] < $test;
        }
        if (in_array($name, ['goldfish', 'pomeranians'])) {
            return COMPOUNDS[$name] > $test;
        }
    }

    return COMPOUNDS[$name] === $test;
}

$sue_one = $sue_two = null;
while (preg_match('/Sue (\d+): (\w+): (\d+), (\w+): (\d+), (\w+): (\d+)/', fgets(STDIN), $matches)) {
    if (compare($matches[2], $matches[3])
        && compare($matches[4], $matches[5])
        && compare($matches[6], $matches[7])) {
        $sue_one = $matches[1];
    }
    if (compare($matches[2], $matches[3], true)
        && compare($matches[4], $matches[5], true)
        && compare($matches[6], $matches[7], true)) {
        $sue_two = $matches[1];
    }
}

fwrite(STDOUT, $sue_one . PHP_EOL);
fwrite(STDOUT, $sue_two . PHP_EOL);